File: backend/services/game/core/models/game_player.py
# Third Party
from django.db import models
from django.utils.translation import gettext_lazy as _

# First Party
from user.models import User

# Local Folder
from .game_player_position import GamePlayerPosition


class GamePlayer(models.Model):
    game = models.ForeignKey(
        to="core.Game",
        on_delete=models.CASCADE,
        related_name="game_players",
        verbose_name=_("Game"),
    )
    user = models.ForeignKey(
        to=User,
        on_delete=models.SET_NULL,
        null=True,
    )
    score = models.IntegerField(default=0, verbose_name=_("Player' score in the game"))
    position = models.IntegerField(choices=GamePlayerPosition, default=0)

    class Meta:
        db_table = "game_player"
        unique_together = [["game", "user"]]

    @property
    def name(self):
        if self.user:
            t = self.game.get_tournament()
            if t:
                u = self.user
                tp = u.tournaments_player.filter(tournament=t, user=self.user).first()
                if tp:
                    return tp.name
            return self.user.username
        return User.anonymous()["username"]

    def to_json(self) -> dict:
        user = self.user
        return {
            "user": user.resume_to_json() if user else User.anonymous(),
            "score": self.score,
            "position": self.position,
        }

    def __str__(self) -> str:
        if self.user:
            return self.user.username
        return User.anonymous()["username"]

File: app_frontend/static/src/js/contracts/game/gamePlayer.js
import { objValidation as validateObj } from "../validation/objValidation.js";


class GameUser {
  static mustHaveKeys = ["id", "username"];

  constructor(id, username, name) {
    this.id = id;
    this.username = username;
    this.name = name;
  }

  static createGameUserFromObj(obj) {
    validateObj(GameUser, obj);
    return new GameUser(obj.id, obj.username, obj.name);
  }

  toJSON() {
    return {
      id: this.id,
      username: this.username,
      name: this.name,
    }
  }
}

class GamePlayer {
  static mustHaveKeys = ["user", "score"];

  constructor(user, score) {
    this.user = GameUser.createGameUserFromObj(user);
    this.score = score;
  }

  static createGamePlayerFromObj(obj) {
    validateObj(GamePlayer, obj);
    validateObj(GameUser, obj.user);
    return new GamePlayer(obj.user, obj.score);
  }

  toJSON() {
    return {
      user: this.user.toJSON(),
      score: this.score,
    }
  }
}

export { GamePlayer };
